crates/nu-engine/src/env.rs

```diff
@@ -157,85 +157,48 @@ pub fn env_to_strings(
-                inner: vec![],
-        })
-    }
+    #[allow(deprecated)]
+    current_dir(engine_state, stack).map(|path| path.to_string_lossy().to_string())
 }
 
-/// Simplified version of current_dir_str() for constant evaluation
+/// Returns the current working directory as a String, which is guaranteed to be canonicalized.
+///
+/// Returns an error if $env.PWD doesn't exist, is not a String, or is not an absolute path.
+#[deprecated(since = "0.92.3", note = "please use `EngineState::cwd()` instead")]
 pub fn current_dir_str_const(working_set: &StateWorkingSet) -> Result<String, ShellError> {
-    if let Some(pwd) = working_set.get_env_var(PWD_ENV) {
-        let span = pwd.span();
-        match pwd {
-            Value::String { val, .. } => {
-                if Path::new(val).is_absolute() {
-                    Ok(val.clone())
-                } else {
-                    Err(ShellError::GenericError {
-                            error: "Invalid current directory".into(),
-                            msg: format!("The 'PWD' environment variable must be set to an absolute path. Found: '{val}'"),
-                            span: Some(span),
-                            help: None,
-                            inner: vec![]
-                    })
-                }
-            }
-            _ => Err(ShellError::GenericError {
-                error: "PWD is not a string".into(),
-                msg: "".into(),
-                span: None,
-                help: Some(
-                    "Cusrrent working directory environment variable 'PWD' must be a string."
-                        .into(),
-                ),
-                inner: vec![],
-            }),
-        }
-    } else {
-        Err(ShellError::GenericError{
-                error: "Current directory not found".into(),
-                msg: "".into(),
-                span: None,
-                help: Some("The environment variable 'PWD' was not found. It is required to define the current directory.".into()),
-                inner: vec![],
-        })
-    }
+    #[allow(deprecated)]
+    current_dir_const(working_set).map(|path| path.to_string_lossy().to_string())
 }
 
-/// Calls current_dir_str() and returns the current directory as a PathBuf
+/// Returns the current working directory, which is guaranteed to be canonicalized.
+/// Unlike `current_dir_const()`, this also considers modifications to the current working directory made on the stack.
+///
+/// Returns an error if $env.PWD doesn't exist, is not a String, or is not an absolute path.
+#[deprecated(since = "0.92.3", note = "please use `EngineState::cwd()` instead")]
 pub fn current_dir(engine_state: &EngineState, stack: &Stack) -> Result<PathBuf, ShellError> {
-    current_dir_str(engine_state, stack).map(PathBuf::from)
+    let cwd = engine_state.cwd(Some(stack))?;
+    canonicalize_with(&cwd, ".").map_err(|_| ShellError::DirectoryNotFound {
```

Comment:
Rather than canonicalizing with `"."`, `fs::canonicalize()` should be called with a comment that we canonicalize a guaranteed absolute path.

Reply:
We can't actually use `fs::canonicalize()` here because we need to call `to_winuser_path()` on Windows. I'd love to use `nu_path::canonicalize()` but that API is not exposed to public. So maybe using `canonicalize_with()` is OK? I'll add a comment explaining this situation.

Reply:
`nu-path::canonicalize()` uses `Path.canonicalize()` which is an alias to `fs::canonicalize()`. What is the problem with `to_winuser_path()`? I'm not familiar with it.

In Nushell, calling `fs::canonicalize()` is OK _after the path has been expanded into an absolute path_`. That's why we always use `nu-path::canonicalize_with()` with the cwd coming from `$env.PWD` which should be guaranteed to be an absolute path, therefore the underlying `fs:canonicalize()` shouldn't use the wrong current dir from `std::env`.

Reply:
On Windows, `fs::canonicalize()` will return UNC paths, which looks like `\\?\C:\Users`. This is considered "canonical", but not very user friendly. `nu-path::canonicalize()` is a wrapper over `fs::canonicalize()` that converts UNC paths into "legacy" paths like `C:\Users` on Windows (on other platforms it does nothing).

Calling `fs::canonicalize()` is incorrect here because we would get UNC paths. That's why I said "I'd love to use `nu-path::canonicalize()`", because that also performs the conversion without supplying a meaningless cwd.

Reply:
Side note: for the "UNC to legacy conversion" functionality, we're currently using the `omnipath` crate. However, the `dunce` crate is a lot more popular, has reached version `1.0`, and offers an API similar to our wrapper. It would be a good idea to use that instead.

Reply:
The reason for the creation of `omnipath` was because the `dunce` crate has issues. E.g.: https://github.com/nushell/nushell/issues/6461#issuecomment-1234793374

If the dunce crate has since fixed them then great! Otherwise using it would be a regression.

Reply:
OK, I see it now in the code, I should have looked properly. Then, yes, let's use `nu-path::canonicalize()`.  The reason why it's not a public API is to not encourage people to use it and use `canonicalize_with()` instead. But we can make it public. Maybe it can be renamed to something slightly worrying, like `canonicalize_unsafe()`, with a docstring to discourage its usage unless you know what you're doing, just to minimize potential misuse.

@ChrisDenton The issue and the linked dunce issue are both closed. Seems fixed now?

Reply:
The nushell issue was closed by switching to omnipath. The dunce issue looks like it was just closed without a fix. Maybe wontfix? I can't tell but it has been years at this point.

Reply:
I tested `dunce` and it doesn't seem to have been fixed.

On the other hand, `omnipath`'s approach is to just strip UNC prefix unconditionally, which is also incorrect. It might be the lesser evil for now. I'll keep looking.

Reply:
This test case fails:

```rust
#[test]
fn long_verbatim_path() {
    let long_filename = str::repeat("A", 500);
    let path = Path::new(r"\\?\C:\").join(long_filename);

    let result = path.to_winuser_path().unwrap();
    // `path` is over the 260 character limit, so we can't simplify it
    assert_eq!(result, path);
}
```

I looked at the code, and indeed `omnipath` tries to use `GetFullPathNameW` to check if stripping prefix is safe. This test case fails because it skips the check when the prefix is a drive letter. I think this is a bug?

Also, while I was testing, I noticed `GetFullPathNameW` was accepting long paths even though I never enabled it on my machine. Would you try this test case and tell me if it works on your machine?

```rust
#[test]
fn GetFullPathNameW_long_path() {
    let long_filename = PathBuf::from(str::repeat("A", 1000));
    let mut buffer = [0u16; 1200];
    let ret = unsafe {
        GetFullPathNameW(
            to_wide(&long_filename).unwrap().as_ptr(),
            1200,
            buffer.as_mut_ptr(),
            ptr::null_mut(),
        )
    };
    // If `GetFullPathNameW` accepts long paths, this should pass.
    assert!(ret >= 1000);
}
```

Reply:
Ah, that's intentional. It's meant for displaying paths to the user. Using a path in filesystem APIs should automatically convert to the `\\?\` form if necessary.

I can add a function that does it only for short paths. Though I'd ask the use case? E.g. [`SetCurrentDirectory`](https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-setcurrentdirectory#remarks) does not support `\\?\` paths. You can technically use such a path but then other APIs break because they assume `SetCurrentDirectory` is using either `C:\` or `\\server\share` style paths (as documented).

In any case it's risky to use paths longer than `MAX_PATH` in `SetCurrentDirectory` because it causes `CreateProcess` to fail, even if long paths are supported. I fairly recently added a warning box about that to the `SetCurrentDirectory` docs.

Reply:
I'm a bit confused. Are you saying the result of `to_winuser_path()` should not be used in Windows APIs? Because I think Nushell is currently doing exactly that.

Reply:
Just for clarification, when exactly does `to_winuser_path()` remove the verbatim prefix?

Reply:
It always tries to. It only doesn't if doing so would change the path such that you couldn't recover the original verbatim path. For example, if the file name is `NUL` or if any component ends with a dot.

It does this by making sure `GetFullPathName` doesn't change the path.